doc/flow-definition.md

```diff
@@ -1,11 +1,308 @@
+* The block return value is not used and no state variable is stored.
+
+> The use of these blocks is encouraged; however, they should only ever be used to store a single, `calculator` state variable (in the first question definition); otherwise they should only be used to store user responses on that `calculator` object.
+
+##### `next_node_calculation(variable_name, &block)`
+
+* These blocks were intended to be used to store state variables that are needed in the `next_node` block.
+* These blocks are executed after all the `on_response` blocks have been executed and before any of the `validate` blocks are executed.
+* The parsed response is passed to the block as the only argument and by convention is named `response`.
+* The block return value is stored on the state object as a state variable named `variable_name`.
+
+> The use of these blocks is deprecated and should never be necessary. Define methods on a `calculator` object instead.
+
+##### `validate(message_key, &block)`
+
+* These blocks are intended to be used to validate the user response.
+* These blocks are executed after all the `next_node_calculation` blocks have been executed and before the `next_node` block is executed.
+* The parsed response is passed to the block as the only argument and by convention is named `response`.
+* If the block return value is truth-y, then no action is taken.
+* If the block return value is false-y, then:
+  * A `SmartAnswer::InvalidResponse` exception is raised with the `message_key` set as the exception message.
+  * Although this exception is handled within the app, it prevents the transition to the next node.
+  * The `message_key` from the exception message is set on the built-in state variable, `error`.
+  * When the question template is re-rendered, the `error` state variable is used to lookup the appropriate validation error message in the [question template](doc/question-templates.md#-error_message-message-).
+
+> The use of these blocks is encouraged. However, they should call `valid_xxx?` methods on the `calculator` state variable and not rely on the `response` argument passed into the block.
+
+##### `next_node(&block)`
+
+* There must only be one of these blocks per question definition.
+* This block is intended to determine which node comes next based on the user responses so far.
+* These blocks are executed after all the `validate` blocks have been executed and before any `save_input_as` blocks are executed.
+* The built-in state variables, `path`, `current_node` & `responses`, are updated if this block returns successfully.
+* The block return value must be the result of calling the `#question` or `#outcome` methods passing in the key of the next node - see the [next node documentation](doc/next-node-rules.md) for more details.
+
+> The use of this block type is *required*. However, it should call methods on the `calculator` state variable and not rely on the `response` argument passed into the block.
+
```

Comment:
No. These are instance methods on an instance of a `XxxCalculator` class. The instance of this class is stored as a state variable, named `calculator`.

An example would be the [`already_getting_maternity_pay?` method][1] which is called from [this `next_node` block][2].

This is something that is explained a bit in the [documentation about new-style flows][3], albeit fairly briefly.

I think it would be better to address your question by having a session where Chris & I explain the new-style flow in person. If necessary we can then update the new-style flow documentation accordingly.

[1]: https://github.com/alphagov/smart-answers/blob/b2fe9c04efdb4efa37f2d0589ef1e536314e0537/lib/smart_answer/calculators/statutory_sick_pay_calculator.rb#L49-L51
[2]: https://github.com/alphagov/smart-answers/blob/b2fe9c04efdb4efa37f2d0589ef1e536314e0537/lib/smart_answer_flows/calculate-statutory-sick-pay.rb#L24
[3]: https://github.com/alphagov/smart-answers/blob/master/doc/refactoring.md